docker-compose-rule-core/src/test/java/com/palantir/docker/compose/execution/DockerComposeShould.java

```diff
@@ -166,37 +172,69 @@ public void throw_exception_when_down_fails_for_a_reason_other_than_the_command_
     @Test
     public void use_the_remove_volumes_flag_when_down_exists() throws IOException, InterruptedException {
         compose.down();
-        verify(executor).execute("down", "--volumes");
+        verify(dockerComposeExecutor).execute("down", "--volumes");
     }
 
     @Test
-    public void parse_the_ps_output_on_ports() throws IOException, InterruptedException {
+    public void get_correct_ports() throws IOException, InterruptedException {
+        when(dockerComposeExecutor.execute(anyVararg())).thenReturn(dockerComposeExecutedProcess);
+        when(dockerComposeExecutedProcess.getInputStream()).thenReturn(toInputStream(CONTAINER_ID));
+        when(dockerComposeExecutedProcess.exitValue()).thenReturn(0);
+
+        when(dockerExecutor.execute(anyVararg())).thenReturn(dockerExecutedProcess);
+        when(dockerExecutedProcess.getInputStream()).thenReturn(toInputStream("0.0.0.0:7000->7000/tcp"));
+        when(dockerExecutedProcess.exitValue()).thenReturn(0);
+
         Ports ports = compose.ports("db");
-        verify(executor).execute("ps", "db");
+        verify(dockerComposeExecutor).execute("ps", "-q", "db");
+        verify(dockerExecutor).execute(
+                "ps", "--no-trunc", "--format", "{{.Ports}}", "--filter", String.format("id=%s", CONTAINER_ID));
         assertThat(ports, is(new Ports(new DockerPort("0.0.0.0", 7000, 7000))));
     }
 
     @Test
-    public void throw_illegal_state_exception_when_there_is_no_container_found_for_ports()
-            throws IOException, InterruptedException {
-        when(executedProcess.getInputStream()).thenReturn(toInputStream(""));
+    public void attempt_to_get_ports_but_service_is_empty() throws IOException, InterruptedException {
         exception.expect(IllegalStateException.class);
-        exception.expectMessage("No container with name 'db' found");
-        compose.ports("db");
+        exception.expectMessage("Service cannot be empty");
+        compose.ports("");
     }
 
     @Test
-    public void pass_concatenated_arguments_to_executor_on_docker_compose_exec()
-            throws IOException, InterruptedException {
-        when(executedProcess.getInputStream()).thenReturn(toInputStream("docker-compose version 1.7.0rc1, build 1ad8866"));
+    public void attempt_to_get_ports_but_service_doesnt_exist() throws IOException, InterruptedException {
+        when(dockerComposeExecutor.execute(anyVararg())).thenReturn(dockerComposeExecutedProcess);
+        when(dockerComposeExecutedProcess.getInputStream()).thenReturn(toInputStream(""));
+        when(dockerComposeExecutedProcess.exitValue()).thenReturn(1);
+
+        exception.expect(IllegalStateException.class);
+        exception.expectMessage("No container ID found for service with name 'service_that_doesnt_exist'.");
+        compose.ports("service_that_doesnt_exist");
+    }
+
+    @Test
+    public void attempt_to_get_ports_but_all_ports_are_open_for_container() throws IOException, InterruptedException {
```

Comment:
Again here there is so much mocking I'm not sure what this is actually testing?

Reply:
Will move to an integration test, but this one checks if you specified no ports for your container. An integration test should make this more clear